[JSC] Remove vtable from AccessCase

Constellation opened the pull request:
#### 1614f1089459294e8a0a478e149c8168283f4607
<pre>
[JSC] Remove vtable from AccessCase
<a href="https://bugs.webkit.org/show_bug.cgi?id=244488">https://bugs.webkit.org/show_bug.cgi?id=244488</a>

Reviewed by Ross Kirsling, Mark Lam and Saam Barati.

This patch leverages std::destroying_delete so that we remove vtable from AccessCase.
AccessCase has m_type which represents derived classes so we can use this value to
dispatch appropriate methods. We can save one vtable pointer, and we can also remove
virtual function call for this class. We also use AccessCase for Replace IC for consistency,
it was using ProxyableAccessCase in some cases, but it is not necessary.

* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::createReplace):
(JSC::AccessCase::fromStructureStubInfo):
(JSC::AccessCase::hasAlternateBaseImpl const):
(JSC::AccessCase::alternateBaseImpl const):
(JSC::AccessCase::cloneImpl const):
(JSC::AccessCase::dump const):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::operator delete):
(JSC::AccessCase::clone const):
(JSC::AccessCase::additionalSet const):
(JSC::AccessCase::hasAlternateBase const):
(JSC::AccessCase::alternateBase const):
(JSC::AccessCase::~AccessCase): Deleted.
* Source/JavaScriptCore/bytecode/AccessCase.h:
(JSC::AccessCase::dumpImpl const):
(JSC::AccessCase::additionalSetImpl const):
(JSC::AccessCase::additionalSet const): Deleted.
* Source/JavaScriptCore/bytecode/GetterSetterAccessCase.cpp:
(JSC::GetterSetterAccessCase::create):
(JSC::GetterSetterAccessCase::cloneImpl const):
(JSC::GetterSetterAccessCase::hasAlternateBaseImpl const):
(JSC::GetterSetterAccessCase::alternateBaseImpl const):
(JSC::GetterSetterAccessCase::~GetterSetterAccessCase): Deleted.
(JSC::GetterSetterAccessCase::clone const): Deleted.
(JSC::GetterSetterAccessCase::hasAlternateBase const): Deleted.
(JSC::GetterSetterAccessCase::alternateBase const): Deleted.
* Source/JavaScriptCore/bytecode/GetterSetterAccessCase.h:
* Source/JavaScriptCore/bytecode/InstanceOfAccessCase.cpp:
(JSC::InstanceOfAccessCase::create):
(JSC::InstanceOfAccessCase::cloneImpl const):
(JSC::InstanceOfAccessCase::clone const): Deleted.
(JSC::InstanceOfAccessCase::~InstanceOfAccessCase): Deleted.
* Source/JavaScriptCore/bytecode/InstanceOfAccessCase.h:
* Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.cpp:
(JSC::IntrinsicGetterAccessCase::cloneImpl const):
(JSC::IntrinsicGetterAccessCase::clone const): Deleted.
* Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.h:
* Source/JavaScriptCore/bytecode/ModuleNamespaceAccessCase.cpp:
(JSC::ModuleNamespaceAccessCase::ModuleNamespaceAccessCase):
(JSC::ModuleNamespaceAccessCase::cloneImpl const):
(JSC::ModuleNamespaceAccessCase::~ModuleNamespaceAccessCase): Deleted.
(JSC::ModuleNamespaceAccessCase::clone const): Deleted.
* Source/JavaScriptCore/bytecode/ModuleNamespaceAccessCase.h:
* Source/JavaScriptCore/bytecode/ProxyableAccessCase.cpp:
(JSC::ProxyableAccessCase::create):
(JSC::ProxyableAccessCase::cloneImpl const):
(JSC::ProxyableAccessCase::~ProxyableAccessCase): Deleted.
(JSC::ProxyableAccessCase::clone const): Deleted.
* Source/JavaScriptCore/bytecode/ProxyableAccessCase.h:
(JSC::ProxyableAccessCase::additionalSetImpl const):
(): Deleted.
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCachePutBy):
* Source/WTF/wtf/FastMalloc.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::NodeRareData::operator delete):
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:

Canonical link: <a href="https://commits.webkit.org/253968@main">https://commits.webkit.org/253968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71dbce146dcbd72e9de3f4675fb13e805c2a327a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87652 "failed 2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31739 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/18407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96856 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150618 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91622 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30097 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79749 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93270 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74410 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24142 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/18407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67157 "Passed tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/79466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27795 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/18407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73171 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27759 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/18407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26036 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2793 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29452 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/37203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76002 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29357 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/18407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16854 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/bytecode/GetterSetterAccessCase.cpp:
```diff
@@ -52,15 +52,7 @@ Ref<AccessCase> GetterSetterAccessCase::create(
     bool viaProxy, WatchpointSet* additionalSet, CodePtr<CustomAccessorPtrTag> customGetter, JSObject* customSlotBase,
     std::optional<DOMAttributeAnnotation> domAttribute, RefPtr<PolyProtoAccessChain>&& prototypeAccessChain)
 {
-    switch (type) {
-    case Getter:
-    case CustomAccessorGetter:
-    case CustomValueGetter:
-        break;
-    default:
-        ASSERT_NOT_REACHED();
-    };
```
What a strange way of writing an assertion this was. 😅

rkirsling commented on Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.h:
```diff
@@ -44,9 +44,9 @@ class IntrinsicGetterAccessCase final : public AccessCase {
+    Ref<AccessCase> cloneImpl() const;
 
-    ~IntrinsicGetterAccessCase() final;
+    ~IntrinsicGetterAccessCase();
```
No `default` for this one?

Constellation replied:
I should add it. Added :)

rkirsling reviewed (approved):
Very fancy! Any idea about the performance impact?

MenloDorian commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -885,7 +885,7 @@ static InlineCacheAction tryCachePutBy(JSGlobalObject* globalObject, CodeBlock*
                 }
 
-                newCase = ProxyableAccessCase::create(vm, codeBlock, AccessCase::Replace, propertyName, slot.cachedOffset(), oldStructure, ObjectPropertyConditionSet(), isProxy);
+                newCase = AccessCase::createReplace(vm, codeBlock, propertyName, slot.cachedOffset(), oldStructure, isProxy);
```
Can you put a comment in the commit message about why this is a bug and the fix here?

saambarati replied:
This was a bug?

Constellation replied:
It was a bug, but no effect on correctness since we are using `virtual` to get a watchpoint-set from ProxyableAccessCase. It is just waste of memory, but that's it.

MenloDorian reviewed (approved):
r=me

saambarati commented on Source/WTF/wtf/FastMalloc.h:
```diff
@@ -338,6 +338,10 @@ using WTF::fastAlignedFree;
         ASSERT(location); \
         return location; \
     } \
+    static void deleteAfterDestruction(void* p) \
```
Should we call this freeAfterDestruction? Delete sorta implies we destroy and free.

saambarati replied:
I guess “AfterDestruction” implies it’s after the  destructor runs. But I still think I’d call this freeAfterDestruction

Constellation replied:
OK, changed it to `freeAfterDestruction`.

Constellation committed:
[JSC] Remove vtable from AccessCase
https://bugs.webkit.org/show_bug.cgi?id=244488

Reviewed by Ross Kirsling, Mark Lam and Saam Barati.

This patch leverages std::destroying_delete so that we remove vtable from AccessCase.
AccessCase has m_type which represents derived classes so we can use this value to
dispatch appropriate methods. We can save one vtable pointer, and we can also remove
virtual function call for this class. We also use AccessCase for Replace IC for consistency,
it was using ProxyableAccessCase in some cases, but it is not necessary.

* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::createReplace):
(JSC::AccessCase::fromStructureStubInfo):
(JSC::AccessCase::hasAlternateBaseImpl const):
(JSC::AccessCase::alternateBaseImpl const):
(JSC::AccessCase::cloneImpl const):
(JSC::AccessCase::dump const):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::operator delete):
(JSC::AccessCase::clone const):
(JSC::AccessCase::additionalSet const):
(JSC::AccessCase::hasAlternateBase const):
(JSC::AccessCase::alternateBase const):
(JSC::AccessCase::~AccessCase): Deleted.
* Source/JavaScriptCore/bytecode/AccessCase.h:
(JSC::AccessCase::dumpImpl const):
(JSC::AccessCase::additionalSetImpl const):
(JSC::AccessCase::additionalSet const): Deleted.
* Source/JavaScriptCore/bytecode/GetterSetterAccessCase.cpp:
(JSC::GetterSetterAccessCase::create):
(JSC::GetterSetterAccessCase::cloneImpl const):
(JSC::GetterSetterAccessCase::hasAlternateBaseImpl const):
(JSC::GetterSetterAccessCase::alternateBaseImpl const):
(JSC::GetterSetterAccessCase::~GetterSetterAccessCase): Deleted.
(JSC::GetterSetterAccessCase::clone const): Deleted.
(JSC::GetterSetterAccessCase::hasAlternateBase const): Deleted.
(JSC::GetterSetterAccessCase::alternateBase const): Deleted.
* Source/JavaScriptCore/bytecode/GetterSetterAccessCase.h:
* Source/JavaScriptCore/bytecode/InstanceOfAccessCase.cpp:
(JSC::InstanceOfAccessCase::create):
(JSC::InstanceOfAccessCase::cloneImpl const):
(JSC::InstanceOfAccessCase::clone const): Deleted.
(JSC::InstanceOfAccessCase::~InstanceOfAccessCase): Deleted.
* Source/JavaScriptCore/bytecode/InstanceOfAccessCase.h:
* Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.cpp:
(JSC::IntrinsicGetterAccessCase::cloneImpl const):
(JSC::IntrinsicGetterAccessCase::clone const): Deleted.
* Source/JavaScriptCore/bytecode/IntrinsicGetterAccessCase.h:
* Source/JavaScriptCore/bytecode/ModuleNamespaceAccessCase.cpp:
(JSC::ModuleNamespaceAccessCase::ModuleNamespaceAccessCase):
(JSC::ModuleNamespaceAccessCase::cloneImpl const):
(JSC::ModuleNamespaceAccessCase::~ModuleNamespaceAccessCase): Deleted.
(JSC::ModuleNamespaceAccessCase::clone const): Deleted.
* Source/JavaScriptCore/bytecode/ModuleNamespaceAccessCase.h:
* Source/JavaScriptCore/bytecode/ProxyableAccessCase.cpp:
(JSC::ProxyableAccessCase::create):
(JSC::ProxyableAccessCase::cloneImpl const):
(JSC::ProxyableAccessCase::~ProxyableAccessCase): Deleted.
(JSC::ProxyableAccessCase::clone const): Deleted.
* Source/JavaScriptCore/bytecode/ProxyableAccessCase.h:
(JSC::ProxyableAccessCase::additionalSetImpl const):
(): Deleted.
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCachePutBy):
* Source/WTF/wtf/FastMalloc.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::NodeRareData::operator delete):
* Source/bmalloc/bmalloc/IsoHeap.h:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:

Canonical link: https://commits.webkit.org/253968@main
